dali/pipeline/workspace/workspace_data_factory.h

```diff
@@ -0,0 +1,318 @@
+bool IsPinned(DeviceWorkspace::output_t<Backend> &t) {
+  return t->is_pinned();
+}
+
+template <typename Backend>
+void Reserve(SupportWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
+  t->reserve(new_num_bytes);
+}
+
+template <typename Backend>
+void Reserve(HostWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
+  for (auto &tensor_ptr : t) {
+    tensor_ptr->reserve(new_num_bytes);
+  }
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+void Reserve(DeviceWorkspace::output_t<Backend> &t, size_t new_num_bytes, int batch_size) {
+  t->reserve(new_num_bytes * batch_size);
+}
+
+template <typename Backend>
+void Reserve(SupportWorkspace::output_t<Backend> &t, size_t new_num_bytes) {
+  t->reserve(new_num_bytes);
+}
+
+template <typename Backend>
+void Reserve(HostWorkspace::output_t<Backend> &t, size_t new_num_bytes) {
+  for (auto &tensor_ptr : t) {
+    tensor_ptr->reserve(new_num_bytes);
+  }
+}
+
+// Device is the same as Mixed
+template <typename Backend>
+bool Reserve(DeviceWorkspace::output_t<Backend> &t, size_t new_num_bytes) {
+  return t->reserve(new_num_bytes);
+}
+
+
+/*
+ * Mappings from DALIOpType, DALITensorDevice to index in storage_owner_t
+ */
+
+constexpr size_t GetStorageIndex(DALIOpType op_type, DALITensorDevice device) {
+  return static_cast<size_t>(op_type) * static_cast<size_t>(DALITensorDevice::COUNT) +
+         static_cast<size_t>(device);
+}
+
+constexpr size_t GetMaxStorageIndex() {
+  return GetStorageIndex(DALIOpType::COUNT, static_cast<DALITensorDevice>(0));
+}
+
+constexpr DALIOpType GetOpType(size_t storage_idx) {
+  return static_cast<DALIOpType>(storage_idx / static_cast<size_t>(DALITensorDevice::COUNT));
+}
+
+constexpr DALITensorDevice GetStorageDevice(size_t storage_idx) {
+  return static_cast<DALITensorDevice>(storage_idx % static_cast<size_t>(DALITensorDevice::COUNT));
+}
+
+// We use a tuple that can hold Output Type from Device, Host, Mixed and Support workspaces,
+// so we have a unifided place that can own any of this type.
+// Additionally, we use order of those types deifned by GetStorageIndex
+// We have 4 workspaces with two possible Backends, obtatining 8 types
+// TODO(klecki): this can be clearer as
+// std::tuple<DeviceOutputType<CPUBackend>, DeviceOutputType<GPUBackend>,
+//            HostOutputType<CPUBackend>, ...
+// but that way we ensure correct order of types and not use 8 static_asserts
+template <int storage_idx>
+struct storage_type_gen {
+  using type = typename workspace_t<GetOpType(
+      storage_idx)>::template output_t<storage_backend_t<GetStorageDevice(storage_idx)>>;
+};
+
+template <int storage_idx>
+using storage_gen_t = typename storage_type_gen<storage_idx>::type;
+
+// using storage_owner_t =
+//     std::tuple<storage_gen_t<0>, storage_gen_t<1>, storage_gen_t<2>, storage_gen_t<3>,
+//                storage_gen_t<4>, storage_gen_t<5>, storage_gen_t<6>, storage_gen_t<7>>;
+
+using storage_owner_t =
```

Comment:
I had to read through a lot of code to grasp what this "storage owner" is. Consider renaming it to something less generic, e.g. `WorkspaceDataStore`.

Reply:
Sure. I'm happy for good type name suggestions. :)

Reply:
I went with tensor_data_store_t as it is more about covering the TensorNodes in graph. Also I want to differentiate from workspace_store_t.